lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,754 @@
+
+  $t->ua->max_redirects(2);
+  $t->get_ok('/1')
+    ->header_is(location => '/4');
+
+  # look at the previous hop
+  is $t->tx->previous->res->headers->location, '/3', 'previous redirect';
+
+  $t->ua->max_redirects(3);
+  $t->get_ok('/1')
+    ->header_is(location => '/5');
+
+  $t->ua->max_redirects(4);
+  $t->get_ok('/1')
+    ->json_is('/message' => "this is five");
+
+When we set C<max_redirects>, it stays set for the life of the test object
+until we change it.
+
+L<Test::Mojo>'s handling of HTTP redirects eliminates the need for making
+many, sometimes an unknown number, of redirections to keep testing precise and
+easy to follow (ahem).
+
+=head2 Cookies and Session Management
+
+We can use L<Test::Mojo> to test applications that keep session state in
+cookies. By default, the L<Mojo::UserAgent> object in L<Test::Mojo> will
+manage session for us by saving and sending cookies automatically, just like
+common web browsers:
+
+  use Test::Mojo;
+  use Test::More;
+
+  my $t = Test::Mojo->new('MyApp');
+
+  # No authorization cookie
+  $t->get_ok('/')
+    ->status_is(401)
+    ->content_is('Please log in');
+
+  # Application sets an authorization cookie
+  $t->post_ok('/login', { password => 'let me in' })
+    ->status_is(200)
+    ->content_is('You are logged in');
+
+  # Sends the cookie from the previous transaction
+  $t->get_ok('/')
+    ->status_is(200)
+    ->content_like(qr/You logged in at \d+/);
+
+  # Clear the cookies
+  $t->reset_session;
+
+  # No authorization cookie again
+  $t->get_ok('/')
+    ->status_is(401)
+    ->content_is('Please log in');
+
+We can also inspect cookies in responses for special values through the
+transaction's response (L<Mojo::Message::Response>) object:
+
+  $t->get_ok('/');
+
+  my ($cookie) = grep { $_->name eq 'smarty' } @{$t->tx->res->cookies};
+  like($cookie && $cookie->to_string, qr(smarty=pants), "cookie found");
+
+=head2 Custom Transactions
+
+Let's say we have an application that responds to a new HTTP verb C<RING> and
+to use it we must also pass in a secret cookie value. This is not a
+problem. We can test the application by creating a L<Mojo::Transaction>
+object, setting the cookie, then passing the transaction object to
+C<request_ok>:
+
+  # Use custom "RING" verb
+  my $tx = $t->ua->build_tx(RING => '/doorbell');
+
+  # Set a special cookie
+  $tx->req->cookies({Secret => "don't tell anybody"});
+
+  # Make the request
+  $t->request_ok($tx)
+    ->status_is(200)
+    ->json_is('/status' => 'ding dong');
+
+=head2 WebSockets
+
+L<Test::Mojo> can also test your WebSocket application. Here are some examples
+from L<Test::Mojo>'s documentation:
+
+  $t->websocket_ok('/time')
+    ->message_ok
+    ->message_like(qr/\d+/)
+    ->message_unlike(qr/\w+/)
+    ->finish_ok;
+
+If our WebSocket sends us JSON messages, L<Test::Mojo> has the ability to
+parse those for us:
+
+  $t->websocket_ok('/hello')
+    ->message_ok
+    ->json_message_like('/howdy' => qr/pard/i);
```

Comment:
This should replace the [cookbook recipe](http://mojolicious.org/perldoc/Mojolicious/Guides/Cookbook#Testing-WebSocket-web-services).